### PR TITLE
ARROW-17991: [Python] [C++] Adding support for IpcWriteOptions to the dataset ipc file writer

### PR DESCRIPTION
**Pull request by joosthooz**
This PR adds the option to pass a populated `IpcWriteOptions` object along with the dataset writer. This allows the user to specify a compression type when writing out .arrow files. 
Note that this makes the dataset ipc writer support more configuration options than the normal ipc writer, so we'll probably follow up on this and make them more consistent again.

### CODE REVIEW

File: python/pyarrow/_dataset.pyx [hunk 1]
```diff
@@ -1174,6 +1189,12 @@ cdef class IpcFileFormat(FileFormat):
 
     def equals(self, IpcFileFormat other):
         return True
+    
+    def make_write_options(self, **kwargs):
+        cdef IpcFileWriteOptions opts = \
+            <IpcFileWriteOptions> FileFormat.make_write_options(self)
+        opts.options = IpcWriteOptions(**kwargs)
```
Comment (jorisvandenbossche) [inline python/pyarrow/_dataset.pyx:1196]: Above, the property is called `write_options`?

**Comment by joosthooz** [comment]
This code seems to work, it is able to create an `ipc::IpcWriteOptions` with the compression field set, and pass it along with the dataset writer. The files written with e.g. lz4 are much smaller, so it seems to actually compress the data. However, there is something weird about the `compression` field (this is a `std::shared_ptr<util::Codec>` that is created in Python). I can set the field, but I cannot read it back (it will just return `None`).
The other fields can be read back. Note that the code that is supposed to return the name of a Code already existed.

File: python/pyarrow/_dataset.pyx (hunk 1 above)
Comment (joosthooz) [reply python/pyarrow/_dataset.pyx]: Thanks, fixed!

**Review by lidavidm — COMMENTED** [review]
Seems reasonable, should we add some unit tests?

**Review by lidavidm — COMMENTED** [review]
LGTM. Just one thing to tweak

**Comment by lidavidm** [comment]
Oh, and it seems the formatter has a diff to apply

File: python/pyarrow/tests/test_dataset.py [hunk 2]
```diff
@@ -3241,6 +3241,40 @@ def test_feather_format(tempdir, dataset_reader):
...
+            write_options = file_format.make_write_options(compression=codec)
+        return
+
+    write_options = file_format.make_write_options(compression=compression)
+    ds.write_dataset(
+        table,
+        str(basedir / "data.arrow"),
+        format=file_format,
+        file_options=write_options
+    )
```
[the remark this answers is not in the document]
Comment (joosthooz) [reply python/pyarrow/tests/test_dataset.py:3271]: Yes, I noticed that too. But as a proxy test I could create a bit larger table with lots of zeroes and check if the filesize is smaller!

File: python/pyarrow/tests/test_dataset.py [hunk 3]
```diff
@@ -3241,6 +3241,40 @@ def test_feather_format(tempdir, dataset_reader):
         dataset_reader.to_table(ds.dataset(basedir, format="feather"))
 
 
+@pytest.mark.pandas
+@pytest.mark.parametrize("compression", [
+    "lz4",
+    "zstd",
+    "brotli"  # not supported
+])
+def test_feather_format_compressed(tempdir, compression, dataset_reader):
```
[the remark this answers is not in the document]
Comment (joosthooz) [reply python/pyarrow/tests/test_dataset.py:3250]: I added a check using `is_available`

**Review by lidavidm — APPROVED** [review]

**Comment by lidavidm** [comment]
@jorisvandenbossche any comments?

**Review by jorisvandenbossche — APPROVED** [review]
Thanks, looking good!

We should add some documentation about it, but that's a general issue for the other formats as well, so that can be done in a follow-up.

**Pull request merged**